src/EditorFeatures/CSharpTest/ForeachToFor/ForeachToForTests.cs

```diff
@@ -0,0 +1,1177 @@
+    {
+        foreach[||](var a in new int[] { 1, 3, 4 })
+        {
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var {|Rename:list|} = new int[] { 1, 3, 4 };
+        for (var {|Rename:i|} = 0; i < list.Length; i++)
+        {
+            var a = list[i];
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task CollectionConflict()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var list = 1;
+
+        foreach[||](var a in new int[] { 1, 3, 4 })
+        {
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var list = 1;
+
+        var {|Rename:list1|} = new int[] { 1, 3, 4 };
+        for (var {|Rename:i|} = 0; i < list1.Length; i++)
+        {
+            var a = list1[i];
+            Console.WriteLine(a);
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task IndexConflict()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach [||] (var a in array)
+        {
+            int i = 0;
+        }
+    }
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        for (var {|Rename:i1|} = 0; i1 < array.Length; i1++)
+        {
+            var a = array[i1];
+            int i = 0;
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task VariableWritten()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach [||] (var a in array)
+        {
+            a = 1;
+        }
+    }
+}
+";
+            await TestMissingInRegularAndScriptAsync(text);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task WrongCaretPosition()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        foreach (var a in array)
+        {
+            [||] 
+        }
+    }
+}
+";
+            await TestMissingInRegularAndScriptAsync(text);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ForeachToFor)]
+        public async Task CaretPosition1()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var array = new int[] { 1, 3, 4 };
+        [||] foreach (var a in array)
```

Comment:
this seems undesirable.  i would require the caret be on the foreach.

Reply:
could be. but will leave it as it is for now.

Reply:
Please do not.  We do not want features liek this to light up in too large a span.  We've done a lot of work to minimize that sort of thing.  For example, this would make the feature show up in a comment on top of the foreach.  That's not desirable.  The feature should appear when the caret intersects 'foreach'.

Optionally, it can also appear inside ```[|foreach (|]``` and also potentially after the close paren of the foreach.  But it should not appear inside the span of the locals/condition/increment.

Reply:
I opened issue tracking this design question. 
https://github.com/dotnet/roslyn/issues/25667